### PR TITLE
MINOR: prefix topics if internal config is set

### PR DESCRIPTION
**Pull request by wcarlson5**
In order to move a topology to another runtime without having to copy over the internal topics it would be good to have the option to not prefix the internal topics with the application ID. So this change will introduce a new config that if set will be the internal topic prefix

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by wcarlson5** [comment]
@ableegoldman This should make it so we can have topics prefixed with whatever is needed

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 1]
```diff
@@ -951,6 +951,8 @@
         public static final String IQ_CONSISTENCY_OFFSET_VECTOR_ENABLED = "__iq.consistency.offset"
             + ".vector.enabled__";
 
+        // Private API used to control the usage of consistency offset vectors
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: This comment seems irrelevant? It's not for the offset vectors right?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java [hunk 2]
```diff
@@ -85,8 +86,13 @@ public void init(final StateStoreContext context, final StateStore root) {
     }
 
     private void initInternal(final InternalProcessorContext<?, ?> context) {
+        final String prefix = StreamsConfig.InternalConfig.getString(
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java:89]: Could we use `ProcessorContextUtils#changelogFor` here as well? Ditto below

**Review by guozhangwang — COMMENTED** [review]
@wcarlson5 just for my understanding: what prefix we would use in practice, in order to allow the topology to be moved between runtimes?

**Comment by guozhangwang** [comment]
Though this comment is not for AK, I'd have to raise to our attention: in ksql we have hard-coded rules to detect internal topics to be cleaned when terminating a query, which is like this today:

```
  private static boolean isInternalTopic(final String topicName, final String applicationId) {
    final boolean prefixMatches = topicName.startsWith(applicationId + "-");
    final boolean suffixMatches = topicName.endsWith(KsqlConstants.STREAMS_CHANGELOG_TOPIC_SUFFIX)
        || topicName.endsWith(KsqlConstants.STREAMS_REPARTITION_TOPIC_SUFFIX)
        || topicName.matches(KsqlConstants.STREAMS_JOIN_REGISTRATION_TOPIC_PATTERN)
        || topicName.matches(KsqlConstants.STREAMS_JOIN_RESPONSE_TOPIC_PATTERN);
    return prefixMatches && suffixMatches;
  }
```

With this change we have to remember updating this function otherwise the cleanup logic would break.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java (hunk 2 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java:89]: It could either be a `StateStoreContext` of a `ProcessorContext` and there are different methods for each. I am not sure why as they seem to do similar things. but I didn't want to make reaching changes here. I can just file a ticket to consolidate those methods into one maybe?

**Comment by wcarlson5** [comment]
@guozhangwang yes that is a good point. When we go to use this we will have to set the config and we can updated the clean up logic then

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java:89]: In the future we would remove the deprecated `init(final ProcessorContext context, final StateStore root)` and then we only need `ProcessorContextUtils#changelogFor(StateStoreContext..)`.

Sounds good for filing a JIRA for consolidating to the `changelogFor`, we can do that when we remove the deprecated `init` function.

**Comment by guozhangwang** [comment]
LGTM! Re-triggering the jenkins tests.

**Comment by guozhangwang** [comment]
@wcarlson5 could you double check if the failed tests are relevant?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java (hunk 2 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java:89]: https://issues.apache.org/jira/browse/KAFKA-13588

**Comment by wcarlson5** [comment]
@guozhangwang Yeah I needed update a few tests. It should be good now.

**Review by guozhangwang — APPROVED** [review]
